database.cc

```diff
@@ -852,6 +852,11 @@ future<> database::parse_system_tables(distributed<service::storage_proxy>& prox
         return do_parse_schema_tables(proxy, db::schema_tables::VIEWS, [this, &proxy, &mm] (schema_result_value_type &v) {
             return create_views_from_schema_partition(proxy, v.second).then([this, &mm] (std::vector<view_ptr> views) {
                 return parallel_for_each(views.begin(), views.end(), [this, &mm] (auto&& v) {
+                    // we fix here the schema in place in oreder to avoid races (write commands comming from other coordinators).
+                    view_ptr fixed_v = maybe_fix_legacy_secondary_index_mv_schema(*this, v, nullptr, preserve_version::yes);
+                    if (fixed_v) {
+                        v = fixed_v;
+                    }
                     return this->add_column_family_and_make_directory(v).then([this, &mm, v] {
                         return maybe_update_legacy_secondary_index_mv_schema(mm.local(), *this, v);
```

Comment:
This second call to `maybe_update_legacy_secondary_index_mv_schema` is a little bit excessive now, right? Since it will internally call `maybe_fix_legacy_secondary_index_mv_schema()`, even though it was possibly just fixed above. I think that in here you just need to pass a simple boolean if the view was fixed or not, and if it was, then call:
```cpp
    if (some_variable_which_indicates_that_the_view_was_fixed_a_couple_of_lines_above) {
        return mm.announce_view_update(view_ptr(fixed_v), true);
    }
```

Reply:
It is not excessive because it creates a different schema (with a new version and not a preserved version). We can still
use the previous call to determine if it is even necessary.
I will change the code to only do the second call if we know for sure the schema needs fixing.

Reply:
Isn't it the case that if view_ptr `v` was  eligible for fixing in `maybe_fix_legacy_secondary_index_mv_schema` a couple of lines above, it will also always need to be changed by `maybe_update_legacy_secondary_index_mv_schema`, but the only difference between `fixed_v` and the output of `maybe_update_legacy_secondary_index_mv_schema` would be a new schema version?

Also, could you remind me one more time why we need to preserve the schema version for this short moment, just to create a new one later?

Reply:
It's not a short moment, this will eventually register in the schema_registry and the other nodes can issue read/write commands on this schema. The schema will be incomplete so we will assert that the base table is not initialized for it or
later on that this schema is not suitable for write.

Reply:
Ok, I think I get it, please confirm - will these two calls result in having two distinct versions of the view schema, one only for requests coming from old nodes, and the new, proper one, which should eventually be used by everyone after the upgrade procedure is over?

Reply:
Yes - exactly :slightly_smiling_face: